### BC Hydro IPP - Resource Analysis
#### Non-Storage Hydro

BC Hydro acquires power from Independent Power Producers (IPPs) to help meet B.C.'s electricity needs.
IPPs include power production companies, municipalities, First Nations and customers.

The analysis is carried out on the most recent [IPP supply list](https://www.bchydro.com/content/dam/BCHydro/customer-portal/documents/corporate/independent-power-producers-calls-for-power/independent-power-producers/ipp-supply-list-in-operation.pdf) and [IPP supply map](https://www.bchydro.com/content/dam/BCHydro/customer-portal/documents/corporate/independent-power-producers-calls-for-power/independent-power-producers/ipp-supply-map.pdf) obtained from [BC Hydro Website](https://www.bchydro.com/work-with-us/selling-clean-energy/meeting-energy-needs/how-power-is-acquired.html).

In [1]:
import pandas as pd
import altair as alt

In [2]:
data = pd.read_csv('IPP__REGION_list.csv')

#### Initial Description of Data


As of April 1, 2025, BC Hydro has 119 Electricity Purchase Agreements (EPAs) with IPPs whose projects achieved commercial operation, as defined in their EPA and are considered to be in operation. In total, these projects are capable of delivering approximately 18,662 gigawatt hours of annual supply and approximately 5,298 megawatts of capacity.

In [3]:
len(data)

119

In [4]:
IPP_Capacity = round(data['Capacity(MW)'].sum(),2)
IPP_Capacity

5298.84

In [5]:
IPP_Energy = round(data['Energy(GWh/yr)'].sum(),2)
IPP_Energy

18662.3

In [6]:
data.replace({'Sub_region':{'CI':'Central Interior',
                               'LM':'Lower Mainland',
                               'VI':'Vancouver Island',
                               'NC':'North Coast',
                               'EK':'East Kootenay',
                               'PR':'Peace River',
                               'KLN':'Kelly Lake Nicola',
                               'RSAC':'Revelstoke Ashton Creek',
                               'SK':'Selkirk'}}, inplace = True)

In [7]:
peace_list = ['Central Interior','Peace River']

In [8]:
columbia_list = ['Revelstoke Ashton Creek', 'Kelly Lake Nicola', 'Mica', 'Selkirk', 'East Kootenay']

In [9]:
data.loc[data['Sub_region'].isin(columbia_list),'Region'] = 'Columbia Region'
data.loc[data['Sub_region'].isin(peace_list),'Region'] = 'Peace Region'
data.loc[data['Sub_region'] == 'Lower Mainland','Region'] = 'Lower Mainland'
data.loc[data['Sub_region'] == 'North Coast','Region'] = 'North Coast'
data.loc[data['Sub_region'] == 'Vancouver Island','Region'] = 'Vancouver Island'

In [10]:
columns = data.columns.to_list()
for column in columns:
    if data[column].dtype == 'object':
        data[column] = data[column].str.strip()

The capacity and GWhr contribution of various resources by Type is given by the data_capacity_type and data_energy_type respectively 

In [11]:
data_capacity_type = data.groupby('Type').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()
data_energy_type = data.groupby('Type').agg({'Energy(GWh/yr)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()

In [12]:
data_energy_type

Type  Energy(GWh/yr)
0  Non-Storage Hydro          7336.4
1      Storage Hydro          4933.3
2            Biomass          2977.6
3               Wind          2161.8
4  Gas Fired Thermal           904.7
5                MSW           166.0
6                ERG           141.0
7             Biogas            37.6
8              Solar             3.9

In [13]:
data_capacity_type

Type  Capacity(MW)
0  Non-Storage Hydro      1983.200
1      Storage Hydro      1292.100
2            Biomass       817.900
3               Wind       747.000
4  Gas Fired Thermal       403.800
5                MSW        24.800
6                ERG        22.800
7             Biogas         5.155
8              Solar         2.090

The type of resource contributing to the largest capacity and energy is Non-Storage Hydro 

#### Non-Storage Hydro 
Non-storage hydro resources are hydroelectric facilities that generate power without relying on large reservoirs to store water. They utilize the natural flow of a river to turn turbines and produce electricity. The output is directly tied to the river’s flow. 

BC Hydro has Electricity Purchase Agreement for Non-Storage Hydro with 66 Independent Power Producers. In total, these 66 projects are capable of delivering approximately 7336.4 gigawatt hours of annual supply and approximately 1983.2 megawatts of capacity. The capacity and GWhr contribution of Non-Storage Hydro in various regions are given by the dataframes energy_NSH, capacity_NSH respectively. 

In [14]:
data_NSH = data[data['Type'] == 'Non-Storage Hydro'].reset_index(drop=True)
energy_NSH = data_NSH.groupby('Region').agg({'Energy(GWh/yr)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_NSH = data_NSH.groupby('Region').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [15]:
energy_NSH

Region  Energy(GWh/yr)
0    Lower Mainland          4215.6
1   Columbia Region          1335.7
2       North Coast          1323.1
3  Vancouver Island           366.2
4      Peace Region            95.8

In [16]:
capacity_NSH

Region  Capacity(MW)
0    Lower Mainland        1046.4
1   Columbia Region         509.1
2       North Coast         313.8
3  Vancouver Island          93.4
4      Peace Region          20.5

For Non-storage Hydro, the capacity and energy contributions of each region individually analysed 

#### 1. Lower Mainland


For the lower mainland region, the capacity and GWhr contribution of Non-Storage Hydro resource by location is given in the capacity_LM_Loc and energy_LM_Loc dataframes respectively.


In [17]:
data_LM = data[(data['Type'] == 'Non-Storage Hydro') & (data['Region']=='Lower Mainland')].reset_index(drop=True)
energy_LM_Loc = data_LM.groupby('Location').agg({'Energy(GWh/yr)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_LM_Loc = data_LM.groupby('Location').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [18]:
capacity_LM_Loc

Location  Capacity(MW)
0           Powell River         258.0
1              Pemberton         189.7
2                Mission         188.7
3               Squamish         172.9
4                Sechelt          91.6
5   Harrison Hot Springs          68.1
6               Whistler          34.0
7            Port Mellon          14.7
8                   Hope          11.0
9             Boston Bar           9.0
10               Agassiz           5.5
11               Laidlaw           3.2

In [19]:
energy_LM_Loc

Location  Energy(GWh/yr)
0           Powell River           889.1
1               Squamish           802.6
2                Mission           793.4
3              Pemberton           716.1
4                Sechelt           447.8
5   Harrison Hot Springs           245.9
6               Whistler           168.0
7            Port Mellon            47.4
8             Boston Bar            37.0
9                   Hope            37.0
10               Agassiz            21.0
11               Laidlaw            10.3

The bar chart energy_LM_Loc_chart, shows for Non-Storage Hydro, the energy contribution of the project and the call process by which it was acquired. 

The bar chart capacity_LM_Loc_chart, shows for Non-Storage Hydro, the capacity contribution of each of these projects and the call process by which it was acquired.


In [42]:
title = alt.TitleParams('The Annual GWhr contribution of Non-Storage Hydro in Lower Mainland by IPP Projects')

capacity_IPP = alt.Chart(data_LM, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Energy(GWh/yr))', title='Energy(GWh/yr)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
energy_LM_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Energy(GWh/yr))',color=alt.value('black'))
energy_LM_Loc_chart

alt.LayerChart(...)

In [43]:
title = alt.TitleParams('The Capacity contribution of Non-Storage Hydro in Lower Mainland by IPP Projects')

capacity_IPP = alt.Chart(data_LM, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Capacity(MW))', title='Capacity(MW)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
cap_LM_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Capacity(MW))',color=alt.value('black'))
cap_LM_Loc_chart

alt.LayerChart(...)

#### 2. Vancouver Island


For the Vancouver Island region, the capacity and GWhr contribution of the Non-Storage Hydro resource by location is given in the capacity_VI_Loc and energy_VI_Loc dataframes respectively.

In [22]:
data_VI = data[(data['Type'] == 'Non-Storage Hydro') & (data['Region']=='Vancouver Island')].reset_index(drop=True)
energy_VI_Loc = data_VI.groupby('Location').agg({'Energy(GWh/yr)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_VI_Loc = data_VI.groupby('Location').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [23]:
energy_VI_Loc

Location  Energy(GWh/yr)
0      Port McNeil           174.9
1     Port Alberni            95.2
2         Ucluelet            47.8
3       Gold River            31.9
4           Tahsis            16.0
5  Gabriola Island             0.4

In [24]:
capacity_VI_Loc

Location  Capacity(MW)
0      Port McNeil          45.0
1     Port Alberni          21.8
2         Ucluelet          15.5
3       Gold River           6.6
4           Tahsis           4.0
5  Gabriola Island           0.5

The bar chart energy_VI_Loc_chart, shows the energy contribution for each of these Non-Storage Hydro projects and the call process by which it was acquired. 

The bar chart capacity_VI_Loc_chart, shows the capacity contribution for each of these Non-Storage Hydro projects and the call process by which it was acquired.


In [44]:
title = alt.TitleParams('The Annual GWhr contribution of Non-Storage Hydro in Vancouver Island by IPP Projects')

capacity_IPP = alt.Chart(data_VI, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Energy(GWh/yr))', title='Energy(GWh/yr)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
energy_VI_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Energy(GWh/yr))',color=alt.value('black'))
energy_VI_Loc_chart

alt.LayerChart(...)

In [45]:
title = alt.TitleParams('The Capacity contribution of Non-Storage Hydro in Vancouver Island by IPP Projects')

capacity_IPP = alt.Chart(data_VI, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Capacity(MW))', title='Capacity(MW)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
cap_VI_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Capacity(MW))',color=alt.value('black'))
cap_VI_Loc_chart

alt.LayerChart(...)

#### 3. Peace Region

For Peace region, the capacity and GWhr contribution of the Non-Storage Hydro resource by location is given in the capacity_PR_Loc and energy_PR_Loc dataframes respectively.

In [27]:
data_PR = data[(data['Type'] == 'Non-Storage Hydro') & (data['Region']=='Peace Region')].reset_index(drop=True)
energy_PR_Loc = data_PR.groupby('Location').agg({'Energy(GWh/yr)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_PR_Loc = data_PR.groupby('Location').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [28]:
energy_PR_Loc

Location  Energy(GWh/yr)
0    McBride            63.3
1  Valemount            32.5

In [29]:
capacity_PR_Loc

Location  Capacity(MW)
0    McBride          12.4
1  Valemount           8.1

The bar chart energy_PR_Loc_chart, shows the energy contribution of the project and the call process by which it was acquired. 

The bar chart capacity_PR_Loc_chart, shows the capacity contribution of each of these projects and the call process by which it was acquired.

In [30]:
title = alt.TitleParams('The Annual GWhr contribution of Non-Storage Hydro in Peace Region by IPP Project')

capacity_IPP = alt.Chart(data_PR, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Energy(GWh/yr))', title='Energy(GWh/yr)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
energy_PR_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Energy(GWh/yr))',color=alt.value('black'))
energy_PR_Loc_chart

alt.LayerChart(...)

In [31]:
title = alt.TitleParams('The Capacity contribution of Non-Storage Hydro in Peace Region by IPP Project')

capacity_IPP = alt.Chart(data_PR, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Capacity(MW))', title='Capacity(MW)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
cap_PR_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Capacity(MW))',color=alt.value('black'))
cap_PR_Loc_chart

alt.LayerChart(...)

#### 4. Columbia Region

For Columbia region, the capacity and GWhr contribution of the Non-Storage Hydro resource by location is given in the capacity_CR_Loc and energy_CR_Loc dataframes respectively.

In [32]:
data_CR = data[(data['Type'] == 'Non-Storage Hydro') & (data['Region']=='Columbia Region')].reset_index(drop=True)
energy_CR_Loc = data_CR.groupby('Location').agg({'Energy(GWh/yr)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_CR_Loc = data_CR.groupby('Location').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [33]:
energy_CR_Loc

Location  Energy(GWh/yr)
0        Trail           627.4
1   Revelstoke           274.7
2       Lytton           223.4
3     Kamloops            81.0
4  Gold Bridge            74.3
5     Lillooet            54.0
6       Sandon             0.9

In [34]:
capacity_CR_Loc

Location  Capacity(MW)
0        Trail         335.0
1   Revelstoke          66.6
2       Lytton          50.0
3  Gold Bridge          21.3
4     Kamloops          20.0
5     Lillooet          16.0
6       Sandon           0.2

The bar chart energy_CR_Loc_chart, shows the energy contribution of the project and the call process by which it was acquired. 

The bar chart capacity_CR_Loc_chart, shows the capacity contribution of each of these projects and the call process by which it was acquired.

In [35]:
title = alt.TitleParams('The Annual GWhr contribution of Non-Storage Hydro in Columbia Region by IPP Project')

capacity_IPP = alt.Chart(data_CR, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Energy(GWh/yr))', title='Energy(GWh/yr)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
energy_CR_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Energy(GWh/yr))',color=alt.value('black'))
energy_CR_Loc_chart

alt.LayerChart(...)

In [36]:
title = alt.TitleParams('The Capacity contribution of Non-Storage Hydro in Columbia Region by IPP Project')

capacity_IPP = alt.Chart(data_CR, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Capacity(MW))', title='Capacity(MW)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
cap_CR_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Capacity(MW))',color=alt.value('black'))
cap_CR_Loc_chart

alt.LayerChart(...)

#### 5. North Coast

For North Coast, the capacity and GWhr contribution of the Non-Storage Hydro resource by location is given in the capacity_NC_Loc and energy_NC_Loc dataframes respectively.

In [37]:
data_NC = data[(data['Type'] == 'Non-Storage Hydro') & (data['Region']=='North Coast')].reset_index(drop=True)
energy_NC_Loc = data_NC.groupby('Location').agg({'Energy(GWh/yr)':'sum'}).sort_values(by = 'Energy(GWh/yr)',ascending = False).reset_index()
capacity_NC_Loc = data_NC.groupby('Location').agg({'Capacity(MW)':'sum'}).sort_values(by = 'Capacity(MW)',ascending = False).reset_index()

In [38]:
energy_NC_Loc

Location  Energy(GWh/yr)
0      Stewart          1229.9
1      Terrace            80.9
2  Bella Bella            12.3

In [39]:
capacity_NC_Loc

Location  Capacity(MW)
0      Stewart         278.8
1      Terrace          20.0
2  Bella Bella          15.0

In [40]:
title = alt.TitleParams('The Annual GWhr contribution of Non-Storage Hydro in North Coast by IPP Project')

capacity_IPP = alt.Chart(data_NC, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Energy(GWh/yr))', title='Energy(GWh/yr)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
energy_NC_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Energy(GWh/yr))',color=alt.value('black'))
energy_NC_Loc_chart

alt.LayerChart(...)

In [41]:
title = alt.TitleParams('The Capacity contribution of Non-Storage Hydro in North Coast by IPP Project')

capacity_IPP = alt.Chart(data_NC, title = title).mark_bar().encode(
                    alt.Y('Project_name', title='Project Name',sort='x'),
                    alt.X('sum(Capacity(MW))', title='Capacity(MW)'),
                    alt.Color('Call_process', title = 'Call Process')).properties( width=300)
cap_NC_Loc_chart = capacity_IPP + capacity_IPP.mark_text(
    color = 'black', align='left', dx = 3).encode(text='sum(Capacity(MW))',color=alt.value('black'))
cap_NC_Loc_chart

alt.LayerChart(...)

#### Conclusion

This study was cariied out to learn about the capacity and energy distribution of Non-Storage Hydro IPP resources around various regions in the province. 